src/bt-hh-io.c

```diff
@@ -0,0 +1,714 @@
+  OPCODE_IDLE_TIME_NTF = 0x84,
+  OPCODE_GET_REPORT_NTF = 0x85,
+  OPCODE_VIRTUAL_UNPLUG_NTF = 0x86
+#if ANDROID_VERSION >= 21
+  ,
+  OPCODE_HANDSHAKE_NTF = 0x87
+#endif
+};
+
+static void (*send_pdu)(struct pdu_wbuf* wbuf);
+static const bthh_interface_t* bthh_interface;
+
+static enum ioresult
+send_ntf_pdu(void* data)
+{
+  /*send notification on I/O thread */
+  if (!send_pdu) {
+    ALOGE("send_pdu is NULL");
+    return IO_OK;
+  }
+  send_pdu(data);
+  return IO_OK;
+}
+
+/*
+ * Protocol helpers
+ */
+
+static long
+read_bthh_hid_info_t(const struct pdu* pdu, unsigned long offset,
+                     bthh_hid_info_t* hid_info)
+{
+  long off = read_pdu_at(pdu, offset, "SCCSSSCSS",
+    hid_info->attr_mask, hid_info->sub_class,  hid_info->app_id,
+    hid_info->vendor_id, hid_info->product_id, hid_info->version,
+    hid_info->ctry_code, hid_info->priority,   hid_info->dl_len);
+  if (off < 0)
+    return -1;
+
+  if (hid_info->dl_len > BTHH_MAX_DSC_LEN) {
+    ALOGE("dl_len is too long");
+    return -1;
+  }
+
+  return read_pdu_at(pdu, off, "m", hid_info->dsc_list, hid_info->dl_len);
+}
+
+static long
+append_bthh_hid_info_t(struct pdu* pdu, const bthh_hid_info_t* hid_info)
+{
+  return append_to_pdu(pdu, "SCCSSSCSSm",
+        (uint16_t)hid_info->attr_mask,  (uint8_t)hid_info->sub_class,
+        (uint8_t)hid_info->app_id,      (uint16_t)hid_info->vendor_id,
+        (uint16_t)hid_info->product_id, (uint16_t)hid_info->version,
+        (uint8_t)hid_info->ctry_code,   (uint16_t)hid_info->priority,
+        (uint16_t)hid_info->dl_len,
+        hid_info->dsc_list, hid_info->dl_len);
+}
+
+/*
+ * Notifications
+ */
+
+static void
+connection_state_cb(bt_bdaddr_t* bd_addr, bthh_connection_state_t state)
+{
+  struct pdu_wbuf* wbuf;
+
+  wbuf = create_pdu_wbuf(6 + /* address */
+                         1, /* connection state */
+                         0, NULL);
+  if (!wbuf)
+    return;
+
+  init_pdu(&wbuf->buf.pdu, SERVICE_BT_HH, OPCODE_CONNECTION_STATE_NTF);
+  if ((append_bt_bdaddr_t(&wbuf->buf.pdu, bd_addr) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)state) < 0))
+    goto cleanup;
+
+  if (run_task(send_ntf_pdu, wbuf) < 0)
+    goto cleanup;
+
+  return;
+cleanup:
+  cleanup_pdu_wbuf(wbuf);
+}
+
+static void
+hid_info_cb(bt_bdaddr_t* bd_addr, bthh_hid_info_t hid_info)
+{
+  struct pdu_wbuf* wbuf;
+
+  wbuf = create_pdu_wbuf(6 + /* address */
+                         27 + hid_info.dl_len, /* hid info */
+                         0, NULL);
+
+  if (!wbuf)
+    return;
+
+  init_pdu(&wbuf->buf.pdu, SERVICE_BT_HH, OPCODE_HID_INFO_NTF);
+  if ((append_bt_bdaddr_t(&wbuf->buf.pdu, bd_addr) < 0) ||
+      (append_bthh_hid_info_t(&wbuf->buf.pdu, &hid_info) < 0))
+    goto cleanup;
+
+  if (run_task(send_ntf_pdu, wbuf) < 0)
+    goto cleanup;
+
+  return;
+cleanup:
+  cleanup_pdu_wbuf(wbuf);
+}
+
+static void
+protocol_mode_cb(bt_bdaddr_t* bd_addr, bthh_status_t hh_status,
+  bthh_protocol_mode_t mode)
```

Comment:
Please align parameters on new lines with the beginning of the other parameters. Here and below.